contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,119 @@
+    ///
+    pub resource Heartbeat : HeartbeatPublic {
+        /// The height at which the first source of randomness was recorded
+        access(self) var initHeight: UInt64?
+        /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+        // TODO: Test at scale
+        access(self) let randomSourceHistory: [[UInt8]]
+
+        init() {
+            self.initHeight = nil
+            self.randomSourceHistory = []
+        }
+
+        /// Getter for the totality of recorded randomness source history
+        ///
+        /// @return An array of random sources, each source an array of UInt8
+        ///
+        pub fun getRandomSourceHistory(): [[UInt8]] {
+            pre {
+                self.initHeight != nil: "Heartbeat has not been initialized"
+            }
+            return self.randomSourceHistory
+        }
+
+        /// Getter for the block height at which the first source of randomness was recorded
+        ///
+        /// @return The block height at which the first source of randomness was recorded
+        ///
+        pub fun getInitHeight(): UInt64 {
+            return self.initHeight ?? panic("Heartbeat has not been initialized")
+        }
+
+        /// Getter for the source of randomness at a given block height. Panics if the requested block height either
+        /// precedes or exceeds the recorded history.
+        ///
+        /// @param atBlockHeight The block height at which to retrieve the source of randomness
+        /// @return The source of randomness at the given block height
+        ///
+        pub fun sourceOfRandomness(atBlockHeight: UInt64): [UInt8] {
+            pre {
+                self.initHeight != nil: "Heartbeat has not been initialized"
+                atBlockHeight >= self.initHeight!: "Requested block height precedes recorded history"
+                atBlockHeight < getCurrentBlock().height: "Source of randomness not yet recorded"
+            }
+            let index = UInt64(atBlockHeight - self.initHeight!)
+
+            assert(
+                index >= 0 && index < UInt64(self.randomSourceHistory.length),
+                message: "Problem finding random source history index"
+            )
+
+            return self.randomSourceHistory[index]
+        }
+
+        /// Callable by owner of the Heartbeat resource, Flow Service Account, records the provided random source
+        ///
+        /// @param randomSourceHistory The random source to record
+        ///
+        pub fun heartbeat(randomSourceHistory: [UInt8]) {
+            let now = getCurrentBlock().height
+            if self.initHeight == nil {
+                self.initHeight = now
+                emit HearbeatInitialized(atBlockHeight: now)
+            }
+            self.randomSourceHistory.append(randomSourceHistory)
+            emit NewSourceOfRandomness(atBlockHeight: now, randomSource: randomSourceHistory)
```

Comment:
I would avoid emitting an event here. It adds no value as this will happen predictably every block.

Reply:
I thought it might be helpful for offchain accounting and in the event someone wanted to gether sor from an indexed sources. @bluesign do you have thoughts on including events in this contract?

Reply:
it can be good to have, I would even think adding SoRA and SoRB together, so we don't need to add it to transaction result etc. @tarakby knows much better, so he can comment here better I guess. 

Also event we can listen with streaming API from execution state, no need to query chain ( constant poll )

Reply:
As an inner protocol developer, I don't use the network much "from outside" to be able to assess the advantage of listening to events Vs querying the chain. But here are some comments, probably all in favor of not emitting events 😅 

- `SoR-A` and `SoR_B` events would be emitted for every block, and in that sense they are expected side effects of executing every block, and not an event users want to be notified about (joining Janez's idea of predictable events). 
- the overhead of two event by block may be not negligible, If a user doesn't want to query for each block, they can  batch-query for every N blocks instead of every block. 
- If we emit events and ask users to track the values, we may not need to have a history array of sources (SoR_B) in the first place. I think the history stored on-chain allows users to query any past value, without having to constantly follow the blocks and listen to all events. If the chain is storing data for you, why would you want to store the same data off-chain (if you are able to listen to events, then you are able to query the chain too).
- I see the advantage of exposing SoR_A, which we discussed in another PR. Still not sure if the best approach is to include it in transaction results (only those calling `revertibleRandom`) or to emit it as an event. But anyway `SoR_A` isn't currently visible to this contract, so it can't be done in this PR.

Reply:
cc @bjartek curious about your thoughts on event emission when source of randomness is committed.

Reply:
I think the events here are ok, there is no more information to add in them really, if we want to have richer event we have to send in metadata that is then emitted, and not sure that is worth it.

Reply:
In your opinion, is there a benefit to including the events if the history is easily queryable? @janezpodhostnik correct me if I'm off, but I think the concern is an outsized performance impact (given the execution every block) by their inclusion for little to no benefit. Is there meaningful reason/use case we should expose the data to offchain systems via events?

Reply:
The performance/size impact overall would be very minimal, but in my opinion so would the information that developers get from it, because not everyone will even be interested in this. 

I also think we can add the events, if there are use cases where the events would be more useful compared to just querying the data.

Reply:
Makes sense, @janezpodhostnik. I've updated the contract to remove the events.